src/Sylius/Component/Grid/Filter/StringFilter.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Grid\Filter;
+
+use Sylius\Component\Grid\Data\DataSourceInterface;
+use Sylius\Component\Grid\Data\ExpressionBuilderInterface;
+use Sylius\Component\Grid\Filtering\FilterInterface;
+
+/**
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ */
+class StringFilter implements FilterInterface
+{
+    const TYPE_EQUAL = 'equal';
+    const TYPE_EMPTY = 'empty';
+    const TYPE_NOT_EMPTY = 'not_empty';
+    const TYPE_CONTAINS = 'contains';
+    const TYPE_NOT_CONTAINS = 'not_contains';
+    const TYPE_STARTS_WITH = 'starts_with';
+    const TYPE_ENDS_WITH = 'ends_with';
+    const TYPE_IN = 'in';
+    const TYPE_NOT_IN = 'not_in';
+
+    /**
+     * {@inheritdoc}
+     */
+    public function apply(DataSourceInterface $dataSource, $name, $data, array $options)
+    {
+        $expressionBuilder = $dataSource->getExpressionBuilder();
+
+        if (!is_array($data)) {
+            $data = array('type' => self::TYPE_CONTAINS, 'value' => $data);
+        }
+
+        $fields = array_key_exists('fields', $options) ? $options['fields'] : array($name);
+
+        $type = $data['type'];
+        $value = array_key_exists('value', $data) ? $data['value'] : null;
+
+        if (1 === count($fields)) {
+            $expression = $this->getExpression($expressionBuilder, $type, $name, $value);
```

Comment:
What about
```php
if (1 === count($fields)) {
    $dataSource->restrict($this->getExpression($expressionBuilder, $type, $name, $value));

    return;
}
```
to avoid nasty ``else`` and nested ``foreach`` in it?